docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,546 @@
+* A ``map<string, float64>`` array has format string ``+m``; its single child
+  has name ``entry`` and format string ``+s``; its two grandchildren have names
+  ``key`` and ``value``, and format strings ``u`` and ``g`` respectively.
+* A ``union<ints: int32, floats: float32>`` with type ids ``4, 5`` has format
+  string ``+u:4,5``; its two children have names ``ints`` and ``floats``,
+  and format strings ``i`` and ``f`` respectively.
+
+
+Structure definitions
+=====================
+
+The following free-standing definitions are enough to support the Arrow
+C data protocol in your project.  Like the rest of the Arrow project, they
+are available under the Apache License.
+
+.. code-block:: c
+
+   #define ARROW_FLAG_ORDERED 1
+   #define ARROW_FLAG_NULLABLE 2
+
+   struct ArrowArray {
+     // Type description
+     const char* format;
+     const char* name;
+     const char* metadata;
+     int32_t flags;
+
+     // Data description
+     int64_t length;
+     int64_t null_count;
+     int64_t offset;
+     int64_t n_buffers;
+     int64_t n_children;
+     const void** buffers;
+     struct ArrowArray** children;
+     struct ArrowArray* dictionary;
+
+     // Release callback
+     void (*release)(struct ArrowArray*);
+     // Opaque producer-specific data
+     void* private_data;
+   };
+
+The ArrowArray structure
+------------------------
+
+.. c:member:: const char* ArrowArray.format
+
+   Mandatory.  A null-terminated, UTF8-encoded string describing
+   the data type.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   Consumers MAY decide not to support all data types, but they
+   should document this limitation.
+
+.. c:member:: const char* ArrowArray.name
+
+   Optional.  A null-terminated, UTF8-encoded string of the field
+   or array name.  This is mainly used to reconstruct child fields
+   of nested arrays.
+
+   Producers MAY decide not to provide this information, and consumers
+   MAY decide to ignore it.  If omitted, MAY be NULL or an empty string.
+
+.. c:member:: const char* ArrowArray.metadata
+
+   Optional.  A null-terminated, UTF8-encoded string describing
+   the type's metadata.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   MUST be a JSON-compatible mapping of UTF8 strings to UTF8 strings.
+   Whitespace MUST only use the 0x20 character. Example::
+
+      {"key1": "base64-encoded value1", "key2": "base64-encode value2"}
+
+   If omitted, MUST be NULL (not an empty string).
+
+   Consumers MAY choose to ignore metadata.  Even then, they SHOULD keep
+   the metadata string around so as to propagate its information to their
+   own consumers.
+
+.. c:member:: int32_t ArrowArray.flags
+
+   Optional.  A bitfield of flags enriching the type or array description.
+   Its value is computed by OR'ing together the flag values.
+   The following flags are available:
```

Comment:
i suppose when added this would be one way of communicating encoding.

Reply:
Hmm, perhaps. We can probably make it required while mentioning that it's gonna be 0 most of the time.